MonoGame.Framework/Graphics/GraphicsResource.cs

```diff
@@ -115,9 +116,19 @@ internal static void DisposeAll()
                 {
                     var target = resource.Target;
                     if (target != null)
+                    {
+                        // Do not dispose another device's graphics resource
+                        if (target is GraphicsResource)
+                        {
+                            if ((target as GraphicsResource).GraphicsDevice != GraphicsDeviceContext.Current.GraphicsDevice)
+                                continue;
+                        }
                         (target as IDisposable).Dispose();
+                    }
+
+                    // We can remove weak handle from collection safely since we have cloned resources dictionary
+                    resources.Remove(resource);
```

Comment:
Yep, I've got it queued up after the `EffectCache` PR. Changing how `GraphicsResource` are disposed of will result in a subtle - and I think harmless - change of behaviour, but we can discuss that in the PR.